core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink,
+) : NSOutputStream(toMemory = Unit) {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    @Volatile
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+        }
+
+    override fun streamStatus() = status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+        }
+    }
+
+    override fun close() {
+        status = NSStreamStatusClosed
+        sink.close()
+    }
+
+    @OptIn(DelicateIoApi::class)
+    override fun write(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        return try {
+            if (isClosed()) throw IOException("Underlying sink is closed.")
+            if (status != NSStreamStatusOpen) return -1
+            status = NSStreamStatusWriting
+            sink.writeToInternalBuffer {
+                it.writeNative(buffer, maxLength.toInt())
+            }
+            status = NSStreamStatusOpen
+            maxLength.convert()
+        } catch (e: Exception) {
+            error = e.toNSError()
+            -1
+        }
+    }
+
+    private fun Buffer.writeNative(source: CPointer<uint8_tVar>?, maxLength: Int) {
+        var currentOffset = 0
+        while (currentOffset < maxLength) {
+            val tail = writableSegment(1)
+
+            val toCopy = minOf(maxLength - currentOffset, Segment.SIZE - tail.limit)
+            tail.data.usePinned {
+                memcpy(it.addressOf(tail.pos), source + currentOffset, toCopy.convert())
+            }
+
+            currentOffset += toCopy
+            tail.limit += toCopy
+        }
+        size += maxLength
+    }
+
+    override fun hasSpaceAvailable() = true
+
+    @OptIn(InternalIoApi::class)
+    override fun propertyForKey(key: NSStreamPropertyKey): Any? = when (key) {
+        NSStreamDataWrittenToMemoryStreamKey -> sink.buffer.readByteArray().toNSData()
```

Comment:
`readByteArray` here will remove data from `buffer` - is it expected behaviour here?

Reply:
Good catch! Yes, comparing to `NSOutputStream.outputStreamToMemory`'s behavior, it should not consume the buffer data, just return its current state as read-only.

Reply:
Now the question is, is there an existing API to get the full `Sink`'s buffer state as a `ByteArray` without consuming it?

Reply:
Ideally this should avoid copying the buffer multiple times as well, so probably need to implement this to natively copy directly to `NSData` anyway.

Reply:
there is `Buffer.copyTo(Buffer)`, so I think similar could be implemented for `NSData`
But for now, it's possible to do `buffer.snapshot()` (will return ByteString), and then via unsafe access get underlying `ByteArray` and create `NSData` from it

Reply:
I think that's the last point which is not done yet, right?

Reply:
I implemented `Buffer.snapshotAsNSData()` to copy the buffer state to `NSData` directly. Yes, this should have been the last feedback item.